launchpad/reward.gno

```diff
@@ -0,0 +1,382 @@
+	}
+
+	depositWithTier, exists := depositsByProject[project.id]
+	if !exists {
+		return project, nil
+	}
+
+	tier30 := project.tiers[30]
+	tier90 := project.tiers[90]
+	tier180 := project.tiers[180]
+
+	// Initialize tier calculations
+	for _, tier := range []struct {
+		id     string
+		amount *u256.Uint
+		start  uint64
+	}{
+		{tier30.id, tier30.tierAmountPerBlockX96.NilToZero(), tier30.started.height},
+		{tier90.id, tier90.tierAmountPerBlockX96.NilToZero(), tier90.started.height},
+		{tier180.id, tier180.tierAmountPerBlockX96.NilToZero(), tier180.started.height},
+	} {
+		if lastCalculateHeightForProjectTier[tier.id] == 0 && !tier.amount.IsZero() {
+			lastCalculateHeightForProjectTier[tier.id] = tier.start
+		}
+	}
+
+	// Calculate rewards for each tier
+	tier30RewardX96, reward30, _ := calculateTierRewards(tier30, height,
+		lastCalculateHeightForProjectTier[tier30.id])
+	tier90RewardX96, reward90, _ := calculateTierRewards(tier90, height,
+		lastCalculateHeightForProjectTier[tier90.id])
+	tier180RewardX96, reward180, _ := calculateTierRewards(tier180, height,
+		lastCalculateHeightForProjectTier[tier180.id])
+
+	// Update project tiers
+	tier30.calculatedAmount += reward30
+	tier90.calculatedAmount += reward90
+	tier180.calculatedAmount += reward180
+
+	// Process deposits for each tier
+	for tierStr, depositIds := range depositWithTier {
+		var (
+			tierAmount uint64
+			rewardX96  *u256.Uint
+		)
+
+		switch tierStr {
+		case "30":
+			tierAmount = tier30.actualDepositAmount
+			rewardX96 = tier30RewardX96
+		case "90":
+			tierAmount = tier90.actualDepositAmount
+			rewardX96 = tier90RewardX96
+		case "180":
+			tierAmount = tier180.actualDepositAmount
+			rewardX96 = tier180RewardX96
+		}
+
+		for _, depositId := range depositIds {
+			deposit := deposits[depositId]
+			updatedDeposit, err := processDepositReward(deposit, rewardX96, tierAmount)
+			if err != nil {
+				continue
+			}
+			deposits[depositId] = updatedDeposit
+		}
+	}
+
+	// Update calculation heights
+	if tier30.started.height != 0 {
+		lastCalculateHeightForProjectTier[tier30.id] = height
+	}
+	if tier90.started.height != 0 {
+		lastCalculateHeightForProjectTier[tier90.id] = height
+	}
+	if tier180.started.height != 0 {
+		lastCalculateHeightForProjectTier[tier180.id] = height
+	}
+
+	return project, nil
+}
+
+func calculateDepositReward() {
+	height := uint64(std.GetHeight())
+
+	if height == lastCalculatedHeight {
+		return
+	}
+	lastCalculatedHeight = height
+	for projectId, project := range projects {
+		updatedProject, err := calculateProjectRewards(project, height)
+		if err != nil {
+			continue
+		}
+		projects[projectId] = updatedProject
+	}
+}
+*/
+// calcDepositRatioX128 calculates the deposit ratio with Q128 precision
+func calcDepositRatioX128(tierAmount uint64, amount uint64) *u256.Uint {
+	amountX128 := new(u256.Uint).Mul(u256.NewUint(amount), q128.Clone())
+	amountX128x := new(u256.Uint).Mul(amountX128, u256.NewUint(1_000_000_000))
```

Comment:
makt this `1_000_000_000` as const value.
it's kinda hard to understand with raw value